exporter/prometheusexporter/accumulator.go

```diff
@@ -327,3 +334,81 @@ func copyMetricMetadata(metric pmetric.Metric) pmetric.Metric {
 
 	return m
 }
+
+func AccumulateNumericValues(prev, current, dest pmetric.NumberDataPoint) {
+	if current.StartTimestamp().AsTime().Before(prev.StartTimestamp().AsTime()) {
+		dest.SetStartTimestamp(current.StartTimestamp())
+	} else {
+		dest.SetStartTimestamp(prev.StartTimestamp())
+	}
+
+	older := prev
+	newer := current
+	if current.Timestamp().AsTime().Before(prev.Timestamp().AsTime()) {
+		older = current
+		newer = prev
+	}
+
+	newer.Attributes().CopyTo(dest.Attributes())
+	dest.SetTimestamp(newer.Timestamp())
+
+	switch newer.ValueType() {
+	case pmetric.NumberDataPointValueTypeDouble:
+		dest.SetDoubleValue(newer.DoubleValue() + older.DoubleValue())
+	case pmetric.NumberDataPointValueTypeInt:
+		dest.SetIntValue(newer.IntValue() + older.IntValue())
+	}
+}
+
+func AccumulateHistogramValues(prev, current, dest pmetric.HistogramDataPoint) {
+	if current.StartTimestamp().AsTime().Before(prev.StartTimestamp().AsTime()) {
+		dest.SetStartTimestamp(current.StartTimestamp())
+	} else {
+		dest.SetStartTimestamp(prev.StartTimestamp())
+	}
+
+	older := prev
+	newer := current
+	if current.Timestamp().AsTime().Before(prev.Timestamp().AsTime()) {
+		older = current
+		newer = prev
+	}
+
+	newer.Attributes().CopyTo(dest.Attributes())
+	dest.SetTimestamp(newer.Timestamp())
```

Comment:
What happens if the `current.Timestamp()` is older but `current.StartTimestamp()` is newer?  Or the inverse?  I'm not sure it's correct to allow these values to be taken from different data points.

Reply:
Do you mean something like this:

```
|---prev.StartTimestamp() ------------------------------------ prev.Timestamp(). ---|
           |---current.StartTimestamp() ---- current.Timestamp(). ---|
```

TBH I'm not sure. We had a similar case for **Sums** and would consider it a single-write principle violation ( https://opentelemetry.io/docs/reference/specification/metrics/data-model/#sums-detecting-alignment-issues)

With the current implementation, we would take the newer one based on the `Timestamp()` comparison. Though here are the options we could consider:

> If the incoming delta time interval has significant overlap with the previous time interval, we assume a violation of the single-writer principle and can be handled with one of the following options:
> - Simply report the inconsistencies in time intervals, as the error condition could be caused by a misconfiguration.
> - Eliminate the overlap / deduplicate on the receiver side.
> - Correct the inconsistent time intervals by differentiating the given Resource and Attribute set used from overlapping time.

Reply:
Do you need any help here?

Reply:
Hi @bitomaxsp, thanks for joining.

The above concern from @Aneurysm9 is what we had left (so far). I don' think I addressed it well. We are discussing about what to do with the case where:

> current.Timestamp() is older but current.StartTimestamp() is newer? Or the inverse?

as I understand, this could happened due to the multi-writers or bugs from the clients.

Reply:
> Do you mean something like this:
> 
> ```
> |---prev.StartTimestamp() ------------------------------------ prev.Timestamp(). ---|
>            |---current.StartTimestamp() ---- current.Timestamp(). ---|
> ```

In this case I think the current implementation would result in `dest` taking `StartTimestamp` from `prev` and `Timestamp` from `current` without any indication that it had combined them.

> TBH I'm not sure. We had a similar case for **Sums** and would consider it a single-write principle violation ( https://opentelemetry.io/docs/reference/specification/metrics/data-model/#sums-detecting-alignment-issues)
> 
> With the current implementation, we would take the newer one based on the `Timestamp()` comparison. Though here are the options we could consider:
> 
> > If the incoming delta time interval has significant overlap with the previous time interval, we assume a violation of the single-writer principle and can be handled with one of the following options:
> > 
> > * Simply report the inconsistencies in time intervals, as the error condition could be caused by a misconfiguration.
> > * Eliminate the overlap / deduplicate on the receiver side.
> > * Correct the inconsistent time intervals by differentiating the given Resource and Attribute set used from overlapping time.

I'm not really sure what the right thing to do here is, but I would suggest at least logging a warning if `current.StartTimestamp().Before(prev.Timestamp())`.  I think this situation would indicate an inconsistency in the production of the metrics and shouldn't be silently accepted.

Reply:
Logging a warning seems reasonable. 
But also should not we drop points that doesn't fit in the time interval (in addition to printing a warning) ?

> > I have a chance to test it a bit. I put to production and it crashes.
> > ```shell
> > panic: runtime error: index out of range [35] with length 35
> > ...
> > ```
> 
> Thanks for testing this out! It must be that this one is missing out the test cases for some parts. I will take a look at your logs and see what I can do.
> 
> Also going to add the warning log as Aneurysm9 commented above

Let me know, i can test it again when you fix it.

Reply:
See my comments below why it's crashing

Reply:
Okay my memory come back a little bit here with the 3 cases:

Case 1: Delta-to-cumulative
```
|prev.StartTimestamp() - prev.Timestamp()|
                                         |current.StartTimestamp() - current.Timestamp()|
|dest.StartTimestamp() ------------------------------------------------ dest.Timestamp()|
```
We take the accumulated sum of them only if the **current.StartTimestamp()** == **prev.StartTimestamp()**

Case 2: Reset

```
|prev.StartTimestamp() - prev.Timestamp()|
                                                   |current.StartTimestamp() - current.Timestamp()|
                                                   |dest.StartTimestamp() ------- dest.Timestamp()|
```

Since the **current.StartTimestamp()** is after the **prev.Timestamp()** - we understand it as the delta has been reset.

Case 3: Anything else - for example:

```
|prev.StartTimestamp() --------------------- prev.Timestamp()|
     |current.StartTimestamp() - current.Timestamp()|
|dest.StartTimestamp() --------------------- dest.Timestamp()|
```
We only keep the latest Timestamp. Which means whatever ended with latest `Timestamp()` will be kept. Which for this case we would log the warning message out.

Reference the Sum approach: https://opentelemetry.io/docs/reference/specification/metrics/data-model/#sums-delta-to-cumulative 

```
Upon receiving future Delta points, we do the following:
- If the next point aligns with the expected next-time window (see [detecting delta restarts](https://opentelemetry.io/docs/reference/specification/metrics/data-model/#sums-detecting-alignment-issues))
  - Update the “last seen” time to align with the time of the current point.
  - Add the current value to the cumulative counter
  - Output a new cumulative point with the original start time and current last seen time and count.
- If the current point precedes the start time, then drop this point. Note: there are algorithms which can deal with late arriving points.
- If the next point does NOT align with the expected next-time window, then reset the counter following the same steps performed as if the current point was the first point seen.
```

---
@bitomaxsp
> I went through original Lev's (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9006) PR. I see that this PR doesn't include Sum accumulation. Did you forget it or removed intentionally? I can help there if needed

I did the delta-to-cumulative for the Sum at the same time Lev made the #9006 as I was in need of using it for the Sum/Counter. So for this PR I left that part out as it has been done.